packages/playout-gateway/src/tsrHandler.ts

```diff
@@ -601,36 +612,60 @@ export class TSRHandler {
 			})),
 			new Promise<UpdateDeviceOperationsResult>((resolve) =>
 				setTimeout(() => {
-					const keys = _.keys(promiseOperations)
+					const keys = Object.keys(promiseOperations)
 					if (keys.length) {
-						this.logger.warn(`Timeout in _updateDevices: ${keys.join(',')}`)
+						this.logger.warn(
+							`Timeout in _updateDevices: ${Object.values(promiseOperations)
+								.map((op) => op.deviceId)
+								.join(',')}`
+						)
 					}
-					resolve({
-						success: false,
-						reason: 'timeout',
-						details: keys,
-					})
+
+					Promise.all(
+						// At this point in time, promiseOperations contains the promises that have timed out.
+						// If we tried to add or re-add a device, that apparently failed so we should remove the device in order to
+						// give it another chance next time _updateDevices() is called.
+						Object.values(promiseOperations)
+							.filter((op) => op.operation === DeviceAction.ADD || op.operation === DeviceAction.READD)
+							.map(async (op) =>
+								// the device was never added, should retry next round
+								this._removeDevice(op.deviceId)
+							)
+					)
+						.catch((e) => {
+							this.logger.error(
+								`Error when trying to remove unsuccessfully initialized devices: ${stringifyIds(
+									Object.values(promiseOperations).map((op) => op.deviceId)
+								)}`,
+								e
+							)
+						})
+						.finally(() => {
+							resolve({
+								success: false,
+								reason: 'error',
+								details: keys,
+							})
+						})
 				}, INIT_TIMEOUT)
 			), // Timeout if not all are resolved within INIT_TIMEOUT
 		])
 
 		await this._reportResult(resultsOrTimeout)
 
-		{
-			const ps: Promise<any>[] = []
-			// Set logDebug on the devices:
-			for (const device of this.tsr.getDevices()) {
-				const deviceOptions = devices.get(device.deviceId)
-				if (deviceOptions) {
-					const debug: boolean = this.getDeviceDebug(deviceOptions)
-					if (device.debugLogging !== debug) {
-						this.logger.info(`Setting logDebug of device ${device.deviceId} to ${debug}`)
-						ps.push(device.setDebugLogging(debug))
-					}
+		const debugLoggingPs: Promise<any>[] = []
+		// Set logDebug on the devices:
+		for (const device of this.tsr.getDevices()) {
+			const deviceOptions = devices.get(device.deviceId)
```

Comment:
It seems weird that we have a variable called `device`, then on an array called `devices` we call get and receives something that isn't an device, but instead what is presumeably a `deviceOptions`.
Would be nice to rename `devices` to `deviceOptions`.

Also, some type declaration would be really appreciated here, since it's not clear what type we actually have

Reply:
Renamed.